Move dictEntry to PM

karpiop opened the pull request:
dictEntry instances will be allocated to PM. Involves only main dictionary entries.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/redis/135)
<!-- Reviewable:end -->

michalbiesek reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @karpiop)

___
*[src/dict.h, line 156 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nrVqS8KyreHrIAHbS:-M-nrVqS8KyreHrIAHbT:b-o9ougg) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L156)):*
> ```
> dictEntry *dictAddRawPM(dict *d, void *key, dictEntry **existing);
> ```
not used anywhere remove


<!-- Sent from Reviewable.io -->

KFilipek reviewed (changes requested):
> Reviewed 7 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: all files reviewed, 10 unresolved discussions (waiting on @karpiop)

___
*[src/db.c, line 180 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nmYAiEQYa_41-tl-n:-M-nmYAiEQYa_41-tl-o:bnuypu4) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/db.c#L180)):*
> ```C
>  * The program is aborted if the key already exists. */
> void dbAdd(redisDb *db, robj *key, robj *val) {
>     sds copy;
> ```

= NULL;
___
*[src/db.c, line 185 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nmZrDDIOOoxNeh7sc:-M-nmZrDDIOOoxNeh7sd:b-jaoc70) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/db.c#L185)):*
> ```C
>     else
>         copy = sdsdup(key->ptr);
>     int retval;
> ```

int retval = 0;
Please don't leave uninitialized values
___
*[src/dict.h, line 155 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHi4P72KQYu6wJxwE:-M-oHi4P72KQYu6wJxwF:bcjhf8b) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L155)):*
> ```objc
> int dictAdd(dict *d, void *key, void *val);
> int dictAddPM(dict *d, void *key, void *val);
> dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing, int dictionary_entries_on_pmem);
> ```

camelCase, same as dictCreate
___
*[src/dict.h, line 157 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHbsp0hHEDvnZJBwD:-M-oHbsp0hHEDvnZJBwE:bcjhf8b) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L157)):*
> ```objc
> dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing, int dictionary_entries_on_pmem);
> dictEntry *dictAddRawPM(dict *d, void *key, dictEntry **existing);
> dictEntry *dictAddOrFind(dict *d, void *key, int dictionary_entries_on_pmem);
> ```

camelCase, same as dictCreate
___
*[src/dict.c, line 292 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHpan3IoGkEQcLK1G:-M-oHpan3IoGkEQcLK1H:bcjhf8b) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L292)):*
> ```C
>  * If key was added, the hash entry is returned to be manipulated by the caller.
>  */
> dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing, int dictionary_entries_on_pmem)
> ```

camelCase, same as dictCreate
___
*[src/dict.c, line 292 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHxf60VIJ2ZXZHpZ7:-M-oHxf60VIJ2ZXZHpZ8:bzqbnp) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L292)):*
> ```C
>  * If key was added, the hash entry is returned to be manipulated by the caller.
>  */
> dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing, int dictionary_entries_on_pmem)
> ```

Maybe const int* to avoid copy creation
___
*[src/dict.c, line 381 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oIFHi1wJ76lp5ygVq:-M-oIFHi1wJ76lp5ygVr:bcjhf8b) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L381)):*
> ```C
>  *
>  * See dictAddRaw() for more information. */
> dictEntry *dictAddOrFind(dict *d, void *key, int dictionary_entries_on_pmem) {
> ```

camelCase, same as dictCreate
___
*[src/dict.c, line 383 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oIFnsEwUs58FS7OEB:-M-oIFnsEwUs58FS7OEC:bcjhf8b) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L383)):*
> ```C
> dictEntry *dictAddOrFind(dict *d, void *key, int dictionary_entries_on_pmem) {
>     dictEntry *entry, *existing;
>     entry = dictAddRaw(d,key,&existing,dictionary_entries_on_pmem);
> ```

camelCase, same as dictCreate
___
*[src/expire.c, line 442 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oIQYGCYNyYzCBrY9K:-M-oIQYGCYNyYzCBrY9L:bj9kldg) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/expire.c#L442)):*
> ```C
>     if (db->id > 63) return;
> 
>     dictEntry *de = dictAddOrFind(slaveKeysWithExpire,key->ptr,server.dictionary_entries_on_pmem);
> ```

variables above have also camelcase naming


<!-- Sent from Reviewable.io -->

karpiop reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: 4 of 7 files reviewed, 10 unresolved discussions (waiting on @karpiop, @KFilipek, and @michalbiesek)

___
*[src/db.c, line 180 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nmYAiEQYa_41-tl-n:-M-ooDHzEb2_W2z1RJBC:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/db.c#L180)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

= NULL;
</blockquote></details>

Done.
___
*[src/db.c, line 185 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nmZrDDIOOoxNeh7sc:-M-ooDsC7ATvG_Vlp3d8:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/db.c#L185)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

int retval = 0;
Please don't leave uninitialized values
</blockquote></details>

Done.
___
*[src/dict.h, line 155 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHi4P72KQYu6wJxwE:-M-ooHQO1ZHch1jHzTFR:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L155)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

camelCase, same as dictCreate
</blockquote></details>

Done.
___
*[src/dict.h, line 156 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nrVqS8KyreHrIAHbS:-M-ooImU6Mi2LBkV8y9B:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L156)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> dictEntry *dictAddRawPM(dict *d, void *key, dictEntry **existing);
> ```
not used anywhere remove
</blockquote></details>

Done.
___
*[src/dict.h, line 157 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHbsp0hHEDvnZJBwD:-M-ooKdy-ksPeeVapay8:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.h#L157)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

camelCase, same as dictCreate
</blockquote></details>

Done.
___
*[src/dict.c, line 292 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oHpan3IoGkEQcLK1G:-M-ooLh-BSpqcSycaFja:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L292)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

camelCase, same as dictCreate
</blockquote></details>

Done.
___
*[src/dict.c, line 381 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oIFHi1wJ76lp5ygVq:-M-ooMjK2vOjv0HpGp5A:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L381)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

camelCase, same as dictCreate
</blockquote></details>

Done.
___
*[src/dict.c, line 383 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-oIFnsEwUs58FS7OEB:-M-ooNOe0KmyUT21B4K4:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/dict.c#L383)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

camelCase, same as dictCreate
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

michalbiesek reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: 4 of 7 files reviewed, 12 unresolved discussions (waiting on @karpiop and @KFilipek)

___
*[src/dict.c, line 334 at r2](https://reviewable.io/reviews/pmem/redis/135#-M-onDgt1qDJHEUcsS4L:-M-onDgt1qDJHEUcsS4M:b5ru8pk) ([raw file](https://github.com/pmem/redis/blob/7a3563253ce489418e84d20022ad2f41e9b4a626/src/dict.c#L334)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> int dictAddPM(dict *d, void *key, void *val)
> {
>     (void)(d);
>     (void)(key);
>     (void)(val);
>     printf("ERROR: dictAddPM is supported only by memkind\n");
>     exit(1);
>      /* unreachable */
>     return NULL;
> }
> ```
</details>

not needed now - remove
this applies also to USE_MEMKIND 
___
*[src/dict.c, line 345 at r2](https://reviewable.io/reviews/pmem/redis/135#-M-onn-J50F8OS5NvBfg:-M-onn-J50F8OS5NvBfh:b-tnyyye) ([raw file](https://github.com/pmem/redis/blob/7a3563253ce489418e84d20022ad2f41e9b4a626/src/dict.c#L345)):*
> ```C
> }
> #endif
> ```

As we spoke offline extend commit message to be more informative - maybe what scenario are covered
___
*[src/t_zset.c, line 2335 at r2](https://reviewable.io/reviews/pmem/redis/135#-M-oo5f53wbcHo8bmoau:-M-oo5f53wbcHo8bmoav:bauxve6) ([raw file](https://github.com/pmem/redis/blob/7a3563253ce489418e84d20022ad2f41e9b4a626/src/t_zset.c#L2335)):*
> ```
> 0
> ```
Maybe some define(s) for  '0' and '1' value ? -  I am not sure what header would be good to place it, but I guess you could figure it out.


<!-- Sent from Reviewable.io -->

karpiop committed:
Move dictEntry to PM

dictEntry instances will be allocated to PM. Involves dictionary entries created using DB API (db.c)

michalbiesek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: 1 of 7 files reviewed, 10 unresolved discussions (waiting on @karpiop and @KFilipek)

___
*[src/db.c, line 185 at r1](https://reviewable.io/reviews/pmem/redis/135#-M-nmZrDDIOOoxNeh7sc:-M-oLcU709PhR5_EbRRB:b-whp7r4) ([raw file](https://github.com/pmem/redis/blob/ef3ae170a51635ac5e5f173b5fa1d46ba7a439a5/src/db.c#L185)):*
<details><summary><i>Previously, karpiop (Paweł Karpiński) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

IMHO  this could be completely avoided using ternary operator:

int retval = (server.dictionary_entries_on_pmem) ?
dictAddPM(db->dict, copy, val) : dictAdd(db->dict, copy, val);

sds copy = (server.keys_on_pmem) ?
sdsdupPM(key->ptr) : sdsdup(key->ptr);


<!-- Sent from Reviewable.io -->

KFilipek reviewed (approved):
> Reviewed 1 of 3 files at r2, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/135)* status: all files reviewed, 1 unresolved discussion (waiting on @karpiop)



<!-- Sent from Reviewable.io -->
